x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -508,6 +518,43 @@ protected void onAbort() {
         context.shutdown();
     }
 
+    protected void callAndResetSaveStateListeners() {
+        Collection<ActionListener<Void>> listeners = saveStateListeners.getAndSet(null);
+        if (listeners != null) {
+            for (ActionListener<Void> l : listeners) {
+                l.onResponse(null);
+            }
+        }
+    }
```

Comment:
This always assumes that the saving state succeeds. If saving the state fails, the listeners are never called.

Reply:
It also seems to me that a previous save state could be getting called (its listener executing), and these listeners get triggered. But, it is not their current state (e.g. the one with "stop at checkpoint") that was saved, but the previous one that was in the middle of saving.

Reply:
good spot! I changed the implementation to get and reset the listener before calling save state. I think that covers it. Now its possible that "stop at checkpoint" is set after the listeners are taken, but this is benign, I added comments in code.